services/webnn/ort/graph_builder_ort.cc

```diff
@@ -248,18 +271,22 @@ std::string GraphBuilderOrt::GetOperandName(uint64_t operand_id) {
   }
 }
 
-template <typename T>
-std::string GraphBuilderOrt::CreateInitializer(base::span<const uint32_t> shape,
-                                               base::span<const T> data,
-                                               OperandDataType data_type) {
+template <typename DataType>
+  requires internal::IsSupportedTensorType<DataType>
+std::string GraphBuilderOrt::CreateInitializer(
+    base::span<const uint32_t> shape,
+    base::span<const DataType> data) {
   std::string name = GetInsertedOperandName(next_operand_id_);
-  OperandInfo operand_info{name, data_type, shape};
+  OperandInfo operand_info{name, TensorTypeMap<DataType>::value, shape};
   base::span<const uint8_t> byte_span;
-  if constexpr (std::floating_point<T>) {
+  if constexpr (std::floating_point<DataType>) {
     // Floating point types do not have unique object representations, but
     // this code appears to be using a byte span to type-erase, which is fine.
     byte_span = base::as_byte_span(base::allow_nonunique_obj, data);
   } else {
+    // It is safe to call `base::as_byte_span(data)` for `base::span<const
+    // Float16>`, because `Float16` is a wrapper around `uint16_t`, not a real
+    // floating point type.
```

Comment:
My concern is `Float16` is a class. The `base::as_byte_span()` of a `std::vector<Float16>` might not give us the same memory layout of `std::vector<uint16_t>`. Or did I miss anything?

Reply:
Oh, good question! I debugged, and find that the `class Float16` type can pass the check `internal::ByteSpanConstructibleFrom<const T&>` at  https://source.chromium.org/chromium/chromium/src/+/main:base/containers/span.h;l=1656;drc=f39c57f31413abcb41d3068cfb2c7a1718003cc5?q=base::as_byte_span&ss=chromium%2Fchromium%2Fsrc, and the webnn wpt test for float16 clamp can pass correctly, it should be OK to use `base::as_byte_span()` for `Float16`.

And I just tried if I put a `std::string` member in the `class Float16`, this check will fail. Anyway, `uint16` member is OK.

Reply:
`Float16` can be converted to byte span, because it has an unique object representation. See `std::has_unique_object_representations_v`:

https://source.chromium.org/chromium/chromium/src/+/main:base/containers/span.h;l=311;drc=f39c57f31413abcb41d3068cfb2c7a1718003cc5?q=base::as_byte_span&ss=chromium%2Fchromium%2Fsrc

https://en.cppreference.com/w/cpp/types/has_unique_object_representations

It doesn't guarantee `Float16`'s memory layout (may be compiler or platform dependent) is the same as `uint16_t` which is what we want.

So, I suppose we should keep using `uint16_t` for float16 data without introducing `Float16` class, WDYT?

Reply:
Agree to keep using uint16_t, I will update the PR, thanks!

Reply:
Updated, PTAL again, thanks!